### PR TITLE
Use refs instead of Vector in Lapack

### PR DESCRIPTION
**Pull request by musm**
Could someone run benchmarks here ? Thanks.

### CODE REVIEW

File: base/linalg/lapack.jl [hunk 1]
```diff
@@ -1355,19 +1357,19 @@ for (gelsd, gelsy, elty, relty) in
                 ccall((@blasfunc($gelsy), liblapack), Void,
                     (Ptr{BlasInt}, Ptr{BlasInt}, Ptr{BlasInt}, Ptr{$elty},
                      Ptr{BlasInt}, Ptr{$elty}, Ptr{BlasInt}, Ptr{BlasInt},
-                     Ptr{$elty}, Ptr{BlasInt}, Ptr{$elty}, Ptr{BlasInt},
```
Comment (musm) [inline base/linalg/lapack.jl]: `Ref` helped me catch this bug/error.  The `ccall` signature here says it's a Ptr to an `$elty` but in fact `rcond` is a `$relty`

**Review by andreasnoack — APPROVED** [review]

File: base/linalg/lapack.jl [hunk 2]
```diff
@@ -1040,7 +1040,7 @@ for (gesvx, elty) in
             ldaf = stride(AF,2)
             nrhs = size(B,2)
             ldb  = stride(B,2)
-            rcond = Vector{$elty}(1)
+            rcond = Ref{$elty}(1)
```
Comment (JeffBezanson) [inline base/linalg/lapack.jl]: For `Vector` the `1` here means to make a 1-element vector, but for `Ref` it initializes the contents to `1`, which is harmless but not necessary. Can use `Ref{$elty}()` instead.

File: base/linalg/lapack.jl [hunk 3]
```diff
@@ -1205,8 +1205,8 @@ for (gelsd, gelsy, elty) in
             end
             newB = [B; zeros($elty, max(0, n - size(B, 1)), size(B, 2))]
             s     = similar(A, $elty, min(m, n))
-            rcond = convert($elty, rcond)
-            rnk   = Vector{BlasInt}(1)
+            rcond = Ref{$elty}(rcond)
```
Comment (alyst) [inline base/linalg/lapack.jl]: AFAIU this one would always be inferred as `::Union{Ref{$elty}, <orig. rcond type>}`, whereas `convert($elty, rcond)` gave `::Union{$elty, <orig. rcond type>}` only when `$elty !== <orig. rcond type>`.
Maybe introduce a new var ("**r**rcond") for a better type inference?

File: base/linalg/lapack.jl [hunk 4]
```diff
@@ -1205,8 +1205,8 @@ for (gelsd, gelsy, elty) in
...
+            rnk   = Ref{BlasInt}()
             info  = Ref{BlasInt}()
             work  = Vector{$elty}(1)
             lwork = BlasInt(-1)
```
Comment (alyst) [inline base/linalg/lapack.jl]: If the goal is to get rid of `Ptr{}`, I guess this one could be converted into `Ref{BlasInt}(-5)` as well.

Comment (andreasnoack) [inline base/linalg/lapack.jl]: `lwork` is only an input parameter. I think it is cleaner only to make the output parameters `Ref`s.

Comment (musm) [inline base/linalg/lapack.jl]: The goal is to get rid of `&x` syntax in `ccall`  not get rid of `Ptr`

Comment (alyst) [inline base/linalg/lapack.jl]: `&` is also used together with `m`, `n` and even expressions like `max(1, stride(A,2))`. Actually, could that be wrong for big-endian Int64 machines?

Comment (musm) [inline base/linalg/lapack.jl]: @Keno has a bit to fix those (since those are trivial changes)

File: base/linalg/lapack.jl (hunk 3 above)
Comment (musm) [inline base/linalg/lapack.jl]: done, now there is no type unstable params looking at `@code_warntype`

**Comment by musm** [comment]
This should now be g2g, can someone trigger nanosolider just to doublecheck?

**Comment by andreasnoack** [comment]
@nanosoldier `runbenchmarks("linalg", vs = ":master")`

**Comment by nanosoldier** [comment]
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/23419#issuecomment-324699083):
```
NanosoldierError: error when preparing/pushing to report repo: failed process: Process(`git push`, ProcessExited(1)) [1]
```
Unfortunately, the logs could not be uploaded.
cc @ararslan

**Comment by musm** [comment]
@ararslan any ideas why the lgo didn't get uploaded?

**Comment by ararslan** [comment]
Looks like some weird Git thing. Let's try again. @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23419#issuecomment-324754457) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e597d77e5592feadce80be5393358536696babcf/105d44a_vs_03c5d7d/report.md). cc @ararslan

**Comment by musm** [comment]
Run again?

**Pull request merged**